### ISSUE TITLE
"Dual pages" option is not available for all document types

### ISSUE THREAD

**Issue by jthornhill** [koreader/koreader#5799]
* KOReader version: v2020.01
* Device: Dell Latitude 7275 (Arch Linux + GNOME + Wayland)

#### Issue
The "Dual Pages" option normally available in the rotation menu under "Orientation" is not always present. I know it is available on mobi and epub filetypes, but it is not available for pdf or cbz (comic book archive) files.

I'm not sure if this is a feature request or a documentation request (if there is some way to enable this option by changing other settings), but it might be worth explaining my use case. Koreader is by far the best program I've found for my Linux + GNOME tablet due to its excellent touchscreen support. The form factor of this device is a 12" widescreen display, and it is ideal for multi-column viewing of these document types.

#### Steps to reproduce
Open documents in the the aforementioned file formats and attempt to activate the Dual Pages option. You will find no such option available.
![image](https://user-images.githubusercontent.com/117894/73419064-b3509b80-4315-11ea-80ae-4fbec376e792.png)

**Comment by Frenzie** [comment]
I'd suspect there's a ticket about this already. But yeah, that's not currently implemented for paged media (e.g., PDF, DjVu, CBZ).

**Comment by pazos** [comment]
ue, sorry. Linked to the wrong issue :p

**Comment by Frenzie** [comment]
Btw, I thought the whole idea behind some… weirder recentish (GNOME 3/Shell) design patterns was touch? Like breaking Alt for menu access. They clearly forgot to pay attention to how Microsoft did that. But that aside.

**Comment by kingo132** [comment]
Want pdf dual pages + 1, especially in landscape view mode of a tall screen device.

**Comment by avalonv** [comment]
Are there plans to implement this? I think this would be a really useful feature, specially with fixed size documents like PDFs...

**Comment by NiLuJe** [comment]
No, this is unlikely to see any traction without outside intervention ;).

**Comment by amitbha** [comment]
Very useful feature on widescreen like PC
Actually, the user experience for .pdf is far less than .epub

**Comment by milanglacier** [comment]
This is a useful feature for me too. We can create bounty for this, can liberapay can create bounty? Or any other platform. If so I am willing to make attribution to the bountyz

**Comment by mergen3107** [comment]
See #10204 and #11085

**Comment by milanglacier** [comment]
> See #10204 and #11085

Thanks for pointing the most recent discussion threads to me.

I can imagine it will be quite a dirty work to implement k2pdfopt in dual page mode. Actually I have never seen any pdf apps (pdf expert, acrobat etc.) have ever implemented reflowed PDF content in dual page mode.

Instead, cropping pages in the "page browser" would be sufficient to me. When navigating the page browser, as long as those white margins being cropped out, it is sufficient enough for 10.3 inch devices.

**Comment by Frenzie** [comment]
> This is a useful feature for me too. We can create bounty for this, can liberapay can create bounty? Or any other platform. If so I am willing to make attribution to the bounty.

I don't think so, unless I overlooked it or they added it more recently. But I don't really see how bounties could affect much of anything here, or put another way it's definitely not a one or two hour proposition.

**Comment by poire-z** [comment]
> Instead, cropping pages in the "page browser" would be sufficient to me. It is also not need to be a complete implementation of cropping, just a full fit to crop out all the margins would be sufficient

Really not the way to go. Page browser is just that: showing the original book pages. No processing, no cropping, no reflow, not what you see on the screen with the Reader app after you tweak the way you want to read it.
Moreover, there are many things that do not (and don't need to) work in Page browser: text selection, highlighting, lookup, following links, TOC, statistics, footer....
We will not make Page browser a reading app because of deficiencies in the main Reading app :) Just fix/improve the main reading app.

**Comment by ava179** [comment]
it's crazy that this issue is still open. dual pages is a must-have feature

**Comment by OGKevin** [comment]
I've spent a couple of hours trying to understand the code base and "reverse engineer" the behavior of the scroll setting(as when scroll/continues setting is enabled, it is technically displaying 2 pages 👀).
I've managed to have a flaky and hardcoded way to display 2 pages, however I'm doing something incorrect with the state management. 

While I try to figure that out, I was wondering if a maintainer could point me in the right direction regarding detecting the device orientation. As this should only be enabled when in landscape, but I can't seem to figure out that part.

So far, this is what I have:

![Image](https://github.com/user-attachments/assets/72d61f1b-e3fa-43bf-83c5-5008e4638674)
![Image](https://github.com/user-attachments/assets/1c22053b-c8f0-4dba-b9d3-29525369bc71)
![Image](https://github.com/user-attachments/assets/dbb452a7-1992-4d46-8242-a317c1e9fe45)

You can see the falkyness with the state, the pages are repeated and the page counter doesn't "skip 2" even though i render 2 pages. 

And here another issue:

![Image](https://github.com/user-attachments/assets/15c7e525-61b8-4c01-9aa3-f2b94b9c6788)

So to summarize my key challenges
- Figure out the device orientation to enable this
- Make it configurable(settings entry)
- Find a way to "center" the panels.
- Figure out the issue with page state management
- Figure out how to align the panels

I would appreciate pointers to solve these issues 👍🏾

**Comment by hius07** [comment]
https://github.com/koreader/koreader/blob/15995650e962ac4cc776237d0f999cb4aed5bf0a/frontend/ui/widget/keyvaluepage.lua#L526-L528

**Comment by OGKevin** [comment]
Ok, I've managed to figure out the state issue and alignment issue. 
So relative page movements and "go to page" works.

I now need to figure out how to "correctly" use the visible area(like the canvas on which the pages are rendered). As just using `Screen:getWith()` isn't sufficient. It did allow for centering.

I need to manage to get the system to update
https://github.com/koreader/koreader/blob/89f7f5ac2d7c279a1ae0d243a564b520ac86cac1/frontend/apps/reader/modules/readerview.lua#L57-L62
As I'm bypassing these values atm, since they position the 2 pages incorrectly. 

---

![Image](https://github.com/user-attachments/assets/8c14ba0f-a1d5-466f-8f78-e4751b4fb54d)

This screenshot shows that the 2 pages are aligned, but as you can see there is left over of the prev page at the bottom, and these new merged panels could be enlarged.

Here example of one that happen to fit the screen almost perfectly:

![Image](https://github.com/user-attachments/assets/b8d87f0e-fde5-4bf2-be17-6cbf81888ae7)

---

As I'm currently just rendering 2 pages:
```
-- Render pages
	for i = #pages, 1, -1 do
		local zoom = sizes[i].zoom

		self.document:drawPage(
			bb,
			x_offset,
			y_offset,
			self.visible_area,
			pages[i],
			zoom,
			self.state.rotation,
			self.state.gamma
		)

		x_offset = x_offset + sizes[i].w
	end
```
I suspect I need to find a way of actually drawing 1 big "picture" and zoom that accordingly to fill the screen.

Nevertheless, I'm still trying to wrap my head around this codebase, hopefully I can have a rough PR soon.

**Comment by NiLuJe** [comment]
Yeah, the `visible_area` stuff takes quite a bit of getting used to to grok...

**Comment by OGKevin** [comment]
> > It also consumes 80GB of memory 😭
> 
> Are you sure you were looking at res and not at virt? :-P

😁 I don't remember, but you might be right.
It's still unusable tho, ended up doing a sparse checkout on the code I need and that makes it happier.

![Image](https://github.com/user-attachments/assets/94626ca6-0980-414e-a130-9f4924422cd2)

---

> Yeah, the `visible_area` stuff takes quite a bit of getting used to to grok...

I'll try to add documentation in the code if I figure it out, and/or add another entry in the docs folder.

---

I've managed to have something working that im satisfied with.

![Image](https://github.com/user-attachments/assets/113f9d38-8bec-463d-9a40-4ce004d764b2)

I rather not make one big PR, so I'll make a bunch of small PR's, and the final one would "enable" the feature. Is that an acceptable approach?

First one would be just the hacked in code turned off by a bool flag that is always false. So I can get some initial feedback, and once I get around adding some ui to toggle the feature on or off we can hook it into the flag that hardcoded to false.

**Comment by Frenzie** [comment]
A separate PR for refactoring could certainly  make sense, but as described it does sound more like a single PR with properly separated commits would serve better.

**Comment by OGKevin** [comment]
Appreciate the kind words ❤! 

You can contribute by trying it out and/or giving feedback!
I am not aware/sure if the maintainers have an easy way to build test builds however, to make testing by non tech folks accessible.

Nevertheless, I will post screenshots on the MR, so if you notice something or think of a nice UX flow, do share :)

**Comment by OGKevin** [comment]
👋🏾 Since the MR didn’t make it upstream, I’ve extracted it into a separate plugin.

It’s still early days, but it seems to work on my machine.
Feel free to give it a test, and please make sure to make backups before you do.

If you encounter any bugs or have feedback, please open an issue. 🙂

https://github.com/OGKevin/comicreader.koplugin
